client/components/voip/User.ts

```diff
@@ -0,0 +1,424 @@
+/**
+ * Class representing SIP UserAgent
+ * @remarks
+ * This class encapsulates all the details of sip.js and exposes
+ * a very simple functions and callback handlers to the outside world.
+ * This class thus abstracts user from Browser specific media details as well as
+ * SIP specific protol details.
+ */
+
+import {
+	UserAgent,
+	UserAgentOptions,
+	UserAgentDelegate,
+	Invitation,
+	InvitationAcceptOptions,
+	Session,
+	SessionState,
+	Registerer,
+} from 'sip.js';
+import { OutgoingRequestDelegate } from 'sip.js/lib/core';
+import { SessionDescriptionHandler } from 'sip.js/lib/platform/web';
+
+import { ICallEventDelegate } from './CallEventDelegate';
+import { CallState } from './Callstate';
+import { IConnectionDelegate } from './ConnectionDelegate';
+import { Operation } from './Operations';
+import { IRegisterHandlerDeligate } from './RegisterHandlerDelegate';
+import Stream from './media/Stream';
+import { Logger } from './utils/Logger';
+// User state is based on whether the User has sent an invite(UAC) or it
+// has received an invite (UAS)
+enum UserState {
+	IDLE,
+	UAC,
+	UAS,
+}
+export class User implements UserAgentDelegate, OutgoingRequestDelegate {
+	private connectionDelegate: IConnectionDelegate;
+
+	private registrationDelegate: IRegisterHandlerDeligate;
+
+	private callEventDelegate: ICallEventDelegate;
+
+	private session: Session | undefined;
+
+	private remoteStream: Stream | undefined;
+
+	private config: any = {};
+
+	userAgentOptions: UserAgentOptions = {};
+
+	userAgent: UserAgent | undefined;
+
+	registerer: Registerer | undefined;
+
+	logger: Logger | undefined;
+
+	private _callState: CallState = CallState.IDLE;
+
+	get callState(): CallState {
+		return this._callState;
+	}
+
+	private _opInProgress: Operation = Operation.OP_NONE;
+
+	get operationInProgress(): Operation {
+		return this._opInProgress;
+	}
+
+	private _userState: UserState | undefined;
+
+	get userState(): any {
+		return this._userState;
+	}
+
+	/* Media Stream functions begin */
+	/** The local media stream. Undefined if call not answered. */
+	get localMediaStream(): MediaStream | undefined {
+		const sdh = this.session?.sessionDescriptionHandler;
+		if (!sdh) {
+			return undefined;
+		}
+		if (!(sdh instanceof SessionDescriptionHandler)) {
+			throw new Error('Session description handler not instance of web SessionDescriptionHandler');
+		}
+		return sdh.localMediaStream;
+	}
+
+	/* Media Stream functions end */
+	constructor(
+		config: any,
+		cDelegate: IConnectionDelegate,
+		rDelegate: IRegisterHandlerDeligate,
+		cEventDelegate: ICallEventDelegate,
+	) {
+		this.config = config;
+		this.connectionDelegate = cDelegate;
+		this.registrationDelegate = rDelegate;
+		this.callEventDelegate = cEventDelegate;
+		this._userState = UserState.IDLE;
+		this.logger = new Logger('User');
```

Comment:
The logger gets initialized with class name. The purpose is that it will help in detecting the exact location of the logs. 
I will change the class name to VoIPUser